src/plugins/filecheck/filecheck.c

```diff
@@ -0,0 +1,365 @@
+		{
+			return i;
+		}
+	}
+	return 0;
+}
+
+static int validateEncoding(const uint8_t *line, iconv_t conv, size_t bytesRead)
+{
+	char *ptr = (char *)line;
+	char outBuffer[LINE_BYTES];
+	char *outPtr = outBuffer;
+	size_t inBytes = bytesRead;
+	size_t outSize = sizeof(outBuffer);
+	int ret = iconv(conv, &ptr, &inBytes, &outPtr, &outSize);
+	if(ret == -1 && errno == EILSEQ)
+	{
+		return ((uint8_t *)ptr - line);
+	}
+	return 0;
+}
+static int checkUnprintable(const uint8_t *line)
+{
+	unsigned int i;
+	for(i = 0; i < elektraStrLen((char *)line); ++i)
+	{
+		if(line[i] < 0x20 || line[i] > 0x7E || line[i] != '\n' || line[i] != '\r')
+			return i;
+	}
+	return 0;
+}
+static long checkFile(Key *parentKey, const char *filename, checkStruct *checkConf)
+{
+	FILE *fp = fopen(filename, "rb");
+	if(fp == NULL)
+	{
+		ELEKTRA_SET_ERRORF(138, parentKey, "Couldn't open file %s", filename);
+		return -1;
+	}
+	iconv_t conv = NULL;
+	if(checkConf->checkEncoding)
+	{
+		if(checkConf->encoding != NULL)
+		{
+			conv = iconv_open(checkConf->encoding, checkConf->encoding);
+		}
+		else
+		{
+			conv = iconv_open("UTF-8", "UTF-8");
+		}
+		if(conv == (iconv_t)(-1))
+		{
+			ELEKTRA_SET_ERRORF(138, parentKey, "Couldn't initialize iconv with encoding %s\n", checkConf->encoding);
+			fclose(fp);
+			return -2;
+		}
+	}
+	uint8_t line[LINE_BYTES];
+	int iconv_ret = 0;
+	int bom_ret = 0;
+	int le_ret = 0;
+	int null_ret = 0;
+	int unprintable_ret = 0;
+	uint8_t firstLine = 1;
+	unsigned long counter = 0;
+	int retVal = 0;
+	if(checkConf->checkLineEnding)
+		validateLineEnding(NULL, NULL, 1);
+	while(!feof(fp))
+	{
+		memset(line, 0, sizeof(line));
+		size_t bytesRead = fread(line, 1, sizeof(line), fp);
+		if(checkConf->checkLineEnding)
+		{
+			le_ret = validateLineEnding(line, &(checkConf->validLE), 0);
+			if(le_ret)
+			{
+				ELEKTRA_SET_ERRORF(137, parentKey, "invalid lineending at position %lu", bytesRead+le_ret);
+				retVal = -1;
+				break;
+			}
+		}
+		if(checkConf->rejectNullByte)
+		{
+			null_ret = checkNull(line, bytesRead);
+			if(null_ret)
+			{
+				ELEKTRA_SET_ERRORF(137, parentKey, "found null-byte at position %lu", bytesRead+null_ret);
+				retVal = -1;
+				break;
+			}		
+		}
+		if(checkConf->checkEncoding)
+		{
+			iconv_ret = validateEncoding(line, conv, bytesRead);
+			if(iconv_ret)
+			{
+				ELEKTRA_SET_ERRORF(137, parentKey, "invalid encoding at position %lu", bytesRead+iconv_ret);
+				retVal = -1;
+				break;
+			}
+		}
+		if(firstLine && checkConf->rejectBom)
+		{
+			bom_ret = checkBom(line);
+			if(bom_ret)
+			{
+				ELEKTRA_SET_ERROR(137, parentKey, "found BOM");
+				retVal = -1;
+				break;
+			}
+			firstLine = 0;
+		}
+		if(checkConf->rejectUnprintable)
+		{
+			unprintable_ret = checkUnprintable(line);
+			if(unprintable_ret)
+			{
+				ELEKTRA_SET_ERRORF(137, parentKey, "unprintable character at position %lu", bytesRead+unprintable_ret);
+				retVal = -1;
+				break;
+			}
+
+		}
+		counter += bytesRead;
+	}
+	if(conv)
```

Comment:
Why is this checked? And why here and not earlier? The error (which is conv == (iconv_t)(-1)) is already checked.